src/gba/cart/gpio.c

```diff
@@ -74,16 +76,22 @@ void GBAHardwareGPIOWrite(struct GBACartridgeHardware* hw, uint32_t address, uin
 	}
 	switch (address) {
 	case GPIO_REG_DATA:
+		hw->writeLatch = value & 0xF;
 		if (!hw->p->vbaBugCompat) {
 			hw->pinState &= ~hw->direction;
-			hw->pinState |= value & hw->direction;
+			hw->pinState |= hw->writeLatch & hw->direction;
 		} else {
-			hw->pinState = value & 0xF;
+			hw->pinState = hw->writeLatch;
 		}
 		_readPins(hw);
 		break;
 	case GPIO_REG_DIRECTION:
 		hw->direction = value & 0xF;
+		if (!hw->p->vbaBugCompat) {
+			hw->pinState &= ~hw->direction;
+			hw->pinState |= hw->writeLatch & hw->direction;
+			_readPins(hw);
```

Comment:
I don't think is this correct. I think the "write" will only affect on writing the data, not the direction.

Reply:
I did test this on my own Emerald cart, changing the direction does in fact change the pin state.

Reply:
Let's say you first just write 0 as the direction (i.e., all in, only external devices are asserting pin state here). You can proceed to write 0xF to data. This doesn't do anything to the pin state yet, as direction does not allow for such. If you proceed to then change the direction to 0xF (i.e., all out, only the CPU (or really, the write latch) is able to assert pin state here), then the pin state becomes 0xF, which external devices can react to.